scipy/stats/stats.py

```diff
@@ -7469,6 +7470,254 @@ def brunnermunzel(x, y, alternative="two-sided", distribution="t",
     return BrunnerMunzelResult(wbfn, p)
 
 
+def _confint_lowerbound(n, quantile, confidence):
+    r"""
+    Compute the lower bound for a one-sided confidence interval
+    for a given
+    - quantile (0<`quantile`<1)
+    - confidence level (0<`confidence`<1)
+    - number of samples `n`.
+
+    Returns the largest index of the sample being a valid lower bound,
+    or `None` if there are not enough samples to derive one.
+
+    Used by the public function confint_quantile().
+
+    .. versionadded:: 1.6.0
+    """
+
+    # compute all probabilities from the binomial distribution for
+    # the quantile of interest
+    bd = binom(n, quantile)
+
+    # the lower bound is the last index before the invert survival
+    # function value for the target confidence level
+    lb = bd.isf(confidence) - 1
+
+    if lb < 0:  # isf returns -1 if there are no matching index
+        return None
+    else:
+        return int(lb)
+
+
+def confint_quantile(x, quantile, confidence, type='one-sided'):
+    r"""Compute non-parametric confidence intervals for any quantile.
+
+    This function implements a non-parametric approach to compute
+    confidence intervals for quantiles. The approach is attributed
+    to Thompson [1]_ and later proven to be applicable to any set
+    of i.i.d. samples [2]_. The computation is based on the observation
+    that the probability of a quantile :math:`q` to be larger than any
+    sample :math:`x_m (1\leq m \leq N)` can be computed as
+
+    .. math::
+
+        \mathbb{P}(x_m \leq q) = 1 - \sum_{k=0}^{m-1} \binom{N}{k}
+        q^k(1-q)^{N-k}
+
+    Furthermore, these probabilities are symmetric, which allows to compute
+    both upper and lower bounds from the same computation:
+
+    .. math::
+
+        \mathbb{P}(x_m \leq q) = \mathbb{P}(x_{N-m+1} \geq 1-q).
+
+    The function computes confidence intervals for a given quantile and
+    confidence level, based on `x` which is either a set of samples
+    (one-dimensional array_like) or the number of samples available.
+    The confidence intervals are valid if and only if the samples are i.i.d.
+
+    Both one-sided and two-sided confidence intervals can be obtained
+    (default is one-sided). The function returns two values: either the bounds
+    for the two one-sided confidence intervals, or the lower and upper bounds
+    of a two-sided confidence interval. The return values are either the
+    indexes of the bounds (if `x` is an integer) or sample values
```

Comment:
"indexes of the bounds" is a little confusing. to me. Does this mean that these would be the indices of `x` ordered samples? 
So one way of using this is to pass in data `x` and get the sample values corresponding with the confidence interval, or I could pass in `len(x)` and use the indices to get the sample values from `sorted(x)`?

Reply:
Yes, that's exactly that. This is meant to offer a more flexible usage of the function, however I do realize it also makes it a bit more confusing... 

I would be fine picking one option only if you think it would be better/clearer. In this case, I would go with input being `len(x)` and output the index(es) of tthe CI. This would emphasis the fact that the numerical values in `x` are irrelevant for this method.

Reply:
I suspect users would prefer the other method, though. So I think it's fine to keep this if it's documented in a way that cannot be misunderstood (within reason). Maybe start 
> if `x` is the set of samples, the return values are the sample values corresponding with the bounds of the confidence interval.

as this is easy to understand. Then, in a separate sentence, you'll have room to explain what it returns if `x` is an integer.

Reply:
That's also what I thought, that's why I included both options in the first place. 

I like your suggestions, will integrate that to clarify the doctring.